src/main/java/htsjdk/samtools/cram/ref/ReferenceContext.java

```diff
@@ -0,0 +1,141 @@
+package htsjdk.samtools.cram.ref;
+
+import htsjdk.samtools.cram.CRAMException;
+import htsjdk.samtools.cram.structure.Slice;
+
+/**
+ * Represents the current state of reference sequence processing.
+ *
+ * Are we handling MULTI_REFERENCE records (-2, from the CRAM spec)?
+ *
+ * Are we handling UNMAPPED_UNPLACED records (-1, from the CRAM spec)?
+ *
+ * Or are we handing a known SINGLE_REFERENCE sequence (0 or higher, from the CRAM spec))?
+ *
+ */
+public class ReferenceContext implements Comparable<ReferenceContext> {
+    // TODO: merge with Slice constants
+    static final int INTERNAL_INT_MULTIPLE = Slice.REF_ID_MULTIPLE; // -2
+    static final int INTERNAL_INT_UNMAPPED = Slice.REF_ID_UNMAPPED; // -1
+
+    public static final ReferenceContext MULTIPLE = new ReferenceContext(INTERNAL_INT_MULTIPLE);
+    public static final ReferenceContext UNMAPPED = new ReferenceContext(INTERNAL_INT_UNMAPPED);
+
+    private final ReferenceContextType type;
+    private final int serializableSequenceId;
+
+    /**
+     * Create a ReferenceContext from a serializable sequence ID
+     * 0 or greater for single reference
+     * -1 for unmapped-unplaced
+     * -2 for multiple reference
+     *
+     * @param serializableSequenceId the sequence ID or sentinel value for constructing this ReferenceContext
+     */
+    public ReferenceContext(final int serializableSequenceId) {
+        this.serializableSequenceId = serializableSequenceId;
+
+        switch (serializableSequenceId) {
+            case INTERNAL_INT_MULTIPLE:
```

Comment:
Since you already have the public static `ReferenceContext` values `MULTIPLE` and `UNMAPPED` anyway, why not reserve this public constructor for only constructing `SINGLE_REFERENCE` contexts, where a legitimate (int >= 0) sequence ID value is required. Then have a separate, private constructor that takes a `ReferenceContextType` for creating the public statics, which can then be treated/referenced as true singletons everywhere they're needed.

Reply:
Agree with the first comment - I'll merge the constants.

Disagree with the second - this check is necessary.  Outside of tests, this constructor is only used by callers which don't know whether it's single-ref or not.  The singletons are used in cases where the caller is aware.